src/azure-cli/azure/cli/command_modules/appconfig/_snapshot_custom_client.py

```diff
@@ -0,0 +1,595 @@
+        if_match=if_match,
+        if_none_match=if_none_match,
+        sync_token=sync_token,
+        **kwargs
+    )
+
+
+def _format_url_section(template, **kwargs):
+    components = template.split("/")
+    while components:
+        try:
+            return template.format(**kwargs)
+        except KeyError as key:
+            formatted_components = template.split("/")
+            components = [c for c in formatted_components if "{}".format(key.args[0]) not in c]
+            template = "/".join(components)
+
+
+class AppConfigSnapshotClient:
+
+    def __init__(self, appConfigClient):
+        appConfiguration = appConfigClient._impl
+        self._serializer = appConfiguration._serialize
+        self._deserializer = appConfiguration._deserialize
+        self._sync_token = appConfiguration._config.sync_token
+        self._endpoint = appConfiguration._config.endpoint
+        self._client = appConfiguration._client
+
+    @classmethod
+    def from_connection_string(cls, connection_string):
+        return cls(AzureAppConfigurationClient.from_connection_string(connection_string))
+
+    def begin_create_snapshot(
+            self,
+            name,
+            filters,
+            composition_type=None,
+            retention_period=None,
+            tags=None,
+            if_match=None,
+            if_none_match=None,
+            **kwargs):
+        """
+        Poll the operation status after a given interval based on the retry-after header (default 5s) to ensure that 
+        the snapshot creation has succeeded or failed.
+        The request times out after 30s by default unless otherwise specified.
+        """
+        timeout = kwargs.pop("timeout", 30)
+        default_polling_interval = kwargs.pop("polling_interval", 5)
+
+
+        from datetime import datetime
+        import time
+        from azure.cli.core.commands.progress import IndeterminateStandardOut
+
+        progress = IndeterminateStandardOut()
+
+        def _get_elapsed_time_from(start_time: datetime):
+            return (datetime.now() - start_time).total_seconds()
+
+        def _delay(sleep_time_seconds):
+            delay_start = datetime.now()
+
+            while _get_elapsed_time_from(delay_start) < sleep_time_seconds:
+                progress.spinner.step(label="Provisioning")
+                time.sleep(1)
+
+        progress.write({"message": "Starting"})
+        self.create_snapshot(
+            name,
+            filters,
+            composition_type,
+            retention_period,
+            tags,
+            if_match=if_match,
+            if_none_match=if_none_match,
+            **kwargs
+        )
+
+        current_state = self._get_snapshot_create_status(name)
+
+        start_time = datetime.now()
+        while current_state.operation_status.status == ProvisioningStatus.RUNNING:
+            progress.spinner.step(label="Provisioning")
+            if _get_elapsed_time_from(start_time) > timeout:
```

Comment:
Looks like this is a 30 second timeout. Seems a bit short. Is there a way for the customer to set timeout? If not, we should at least allow a long one like ten minutes at least. But, I don't see the point for a built-in timeout on the client side as long as the operation is still running.